### PR TITLE
Rework MapperService merge methods

### PR DESCRIPTION
**Pull request by javanna**
The different merge methods on MapperService perform at their core two tasks:
1) merge the new mappings with the current ones
2) validate and commit the resulting document mapper

It is beneficial to split this into two separate steps, which can be used separately.

This will help when removing the node mapping refresh action as a follow-up, as it will be possible to merge incoming mappings without applying them.

Also, we are decreasing the usages of DocumentMapper#merge which will be removed as a follow-up, and isolating the pure merging functionality so that merging can be extracted and performed without requiring a full-blown mapper service.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 1]
```diff
@@ -261,65 +261,58 @@ private void assertMappingVersion(
...
-        // commit the change
         this.mapper = newMapper;
         assert assertSerialization(newMapper);
+        return newMapper;
+    }
 
+    private DocumentMapper newDocumentMapper(Mapping mapping, MergeReason reason) {
+        DocumentMapper newMapper = new DocumentMapper(indexSettings, indexAnalyzers, documentParser, mapping);
+        newMapper.root().fixRedundantIncludes();
+        newMapper.validate(indexSettings, reason != MergeReason.MAPPING_RECOVERY);
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:294]: It's a bit annoying that we have to call these two methods after the mapper has been created - maybe we can move them into the DocumentMapper constructor in a follow-up?

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 2]
```diff
@@ -261,65 +261,58 @@ private void assertMappingVersion(
...
         return newMapper;
     }
 
+    private synchronized Mapping mergeMappings(String mappingType, CompressedXContent mappingSource, MergeReason reason) {
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:298]: I don't think this needs to be `synchronized`, given that it's only called from `mergeAndApplyMappings()` which is already `synchronized`?

**Review by romseygeek — APPROVED** [review]
LGTM! This makes things much easier to follow, thanks.

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:294]: I was going to leave exactly the same question: I see that there are places where we only create the mapper without validating it, which is odd. I am going to have to dig on that.

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 2 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:298]: that may totally be a leftover as the previous internalMerge were all synchronized

**Pull request merged**